nodejs/eks/nodes/instances.ts

```diff
@@ -0,0 +1,209 @@
+    const zone = pulumi.all([placementGroupAvailabilityZone, supportedAZs]).apply(([zone, azs]) => {
+        if (azs.includes(zone)) {
+            return zone;
+        } else {
+            throw new pulumi.InputPropertyError({
+                propertyPath: "placementGroupAvailabilityZone",
+                reason: `The provided availability zone ('${zone}') is not supported by the instance type. Supported AZs: [${azs.join(
+                    ", ",
+                )}]`,
+            });
+        }
+    });
+
+    const filteredSubnets = aws.ec2.getSubnetsOutput(
+        {
+            filters: [
+                {
+                    name: "availability-zone",
+                    values: [zone],
+                },
+                {
+                    name: "subnet-id",
+                    values: subnetIds,
+                },
+            ],
+        },
+        opts,
+    );
+
+    return pulumi.all([filteredSubnets, supportedAZs]).apply(([subnets, azs]) => {
+        if (subnets.ids.length === 0) {
+            throw new pulumi.InputPropertyError({
+                propertyPath: "placementGroupAvailabilityZone",
+                reason: `None of the configured subnets are in the provided availability zone ('${zone}'). Choose a different availability zone or change the subnets. Supported AZs: [${azs.join(
+                    ", ",
+                )}]`,
+            });
+        }
+
+        return subnets.ids;
+    });
+}
+
+/**
+ * Gets the availability zones that support a specific EC2 instance type.
+ * @param instanceType The EC2 instance type to check availability for (e.g. "t3.micro")
+ * @param opts Options for the underlying invokes used to retrieve the information
+ * @returns A list of availability zone IDs where the instance type is available
+ */
+function getSupportedAZs(
+    instanceType: pulumi.Input<string>,
+    opts: pulumi.InvokeOutputOptions,
+): pulumi.Output<string[]> {
+    return aws.ec2.getInstanceTypeOfferingsOutput(
+        {
+            filters: [
+                {
+                    name: "instance-type",
+                    values: [instanceType],
+                },
+            ],
+            locationType: "availability-zone",
+        },
+        opts,
+    ).locations;
+}
+
+/**
+ * Gets the first instance type from a list of instance types, or returns a default instance type if none are provided.
+ * This is used to determine which instance type is configured for EFA.
+ *
+ * @param instanceTypes - list of EC2 instance types
+ * @returns The first instance type from the list, or "t3.medium" if no instance types are provided
+ */
+function getEfaInstanceType(
+    instanceTypes: pulumi.Input<pulumi.Input<string>[]> | undefined,
+): pulumi.Output<string> {
+    return instanceTypes
+        ? pulumi.output(instanceTypes).apply((types) => {
+              return types.length > 0 ? types[0] : DEFAULT_INSTANCE_TYPE;
```

Comment:
The other option here would be enforcing that EFA only supports homogenous node groups. IMO that's too limiting, users might wanna choose different types of similar instance types to ensure availability of those.